File: python/tempmail_wrapper/providers/guerrilla_mail.py
"""guerrillamail.com provider implementation."""

from __future__ import annotations

from datetime import datetime

from ..base import TempMailProvider
from ..exceptions import NotFoundError, RateLimitError, TempMailError
from ..http_client import HttpClient
from ..models import Message, MessageDetail


class GuerrillaMailProvider(TempMailProvider):
    """Provider for guerrillamail.com — uses session cookie."""

    BASE_URL = "https://api.guerrillamail.com/ajax.php"

    def __init__(
        self,
        proxies: list[str] | None = None,
        random_ua: bool = True,
        **kwargs,
    ) -> None:
        self._http = HttpClient(proxies=proxies, random_ua=random_ua, use_cookies=True)
        self._email: str | None = None

    @property
    def name(self) -> str:
        return "guerrillamail"

    def _request(self, func: str, params: dict[str, str] | None = None) -> dict:
        params = params or {}
        params["f"] = func
        resp = self._http.get(self.BASE_URL, params=params, timeout=30)
        if resp.status_code == 429:
            raise RateLimitError("Rate limit exceeded", self.name)
        if resp.status_code != 200:
            raise TempMailError(f"API error: {resp.status_code}", self.name)
        return resp.json()

    def generate_email(self) -> str:
        data = self._request("get_email_address")
        self._email = data.get("email_addr")
        if not self._email:
            raise TempMailError("Failed to generate email", self.name)
        return self._email

    def get_inbox(self, email: str) -> list[Message]:
        data = self._request("get_email_list", {"offset": "0"})
        messages = []
        for item in data.get("list", []):
            msg = Message(
                id=str(item["mail_id"]),
                sender=item.get("mail_from", ""),
                subject=item.get("mail_subject", ""),
                date=datetime.fromtimestamp(int(item.get("mail_timestamp", 0))),
            )
            messages.append(msg)
        return messages

    def read_message(self, message_id: str) -> MessageDetail:
        data = self._request("fetch_email", {"email_id": message_id})
        if not data or "error" in data:
            raise NotFoundError(f"Message {message_id} not found", self.name)

        attachments = []
        if isinstance(data.get("attachments"), list):
            for att in data["attachments"]:
                attachments.append({
                    "filename": att.get("filename", ""),
                    "content_type": att.get("content_type", ""),
                    "size": att.get("size", 0),
                })

        return MessageDetail(
            id=str(data.get("mail_id", message_id)),
            sender=data.get("mail_from", ""),
            subject=data.get("mail_subject", ""),
            date=datetime.fromtimestamp(int(data.get("mail_timestamp", 0))),
            body_text=data.get("mail_body", ""),
            body_html=data.get("mail_body_html", ""),
            attachments=attachments,
        )

    def delete_email(self, email: str) -> bool:
        self._http = HttpClient(
            proxies=self._http.proxies,
            random_ua=self._http.random_ua,
            use_cookies=True,
        )
        self._email = None
        return True

File: rust/src/providers/guerrilla_mail.rs
use async_trait::async_trait;
use serde::Deserialize;

use crate::error::TempMailError;
use crate::httpclient::HttpClient;
use crate::models::{Message, MessageDetail};
use crate::providers::TempMailProvider;

const BASE_URL: &str = "https://api.guerrillamail.com/ajax.php";

/// Configuration for guerrillamail.com provider.
#[derive(Debug, Clone, Default)]
pub struct GuerrillaMailConfig {
    /// Preferred language (default: "en").
    pub lang: String,
}

#[derive(Debug, Clone)]
pub struct GuerrillaMail {
    client: HttpClient,
    config: GuerrillaMailConfig,
    session_id: std::sync::Arc<tokio::sync::Mutex<Option<String>>>,
}

impl GuerrillaMail {
    pub fn new(client: HttpClient, config: Option<GuerrillaMailConfig>) -> Self {
        Self {
            client,
            config: config.unwrap_or(GuerrillaMailConfig {
                lang: "en".to_string(),
            }),
            session_id: std::sync::Arc::new(tokio::sync::Mutex::new(None)),
        }
    }
}

#[derive(Deserialize)]
struct EmailAddrResponse {
    email_addr: String,
    sid_token: String,
}

#[derive(Deserialize)]
struct InboxResponse {
    list: Option<Vec<RawGuerrillaMessage>>,
}

#[derive(Deserialize)]
struct RawGuerrillaMessage {
    mail_id: String,
    mail_from: String,
    mail_subject: String,
    mail_date: String,
    mail_excerpt: Option<String>,
    mail_body: Option<String>,
    content_type: Option<String>,
}

#[derive(Deserialize)]
struct EmailBodyResponse {
    mail_id: String,
    mail_from: String,
    mail_subject: String,
    mail_date: String,
    mail_body: Option<String>,
    content_type: Option<String>,
}

fn parse_date(s: &str) -> chrono::DateTime<chrono::Utc> {
    chrono::NaiveDateTime::parse_from_str(s, "%Y-%m-%d %H:%M:%S")
        .map(|dt| dt.and_utc())
        .unwrap_or_else(|_| chrono::Utc::now())
}

#[async_trait]
impl TempMailProvider for GuerrillaMail {
    async fn generate_email(&self) -> Result<String, TempMailError> {
        let resp: EmailAddrResponse = self
            .client
            .get_json(&format!(
                "{}?f=get_email_address&lang={}",
                BASE_URL, self.config.lang
            ))
            .await?;

        let mut sid = self.session_id.lock().await;
        *sid = Some(resp.sid_token);

        Ok(resp.email_addr)
    }

    async fn get_inbox(&self, _email: &str) -> Result<Vec<Message>, TempMailError> {
        let sid = self.session_id.lock().await;
        let sid = sid
            .as_ref()
            .ok_or_else(|| TempMailError::Api("Must call generate_email first".into()))?
            .clone();

        let inbox: InboxResponse = self
            .client
            .get_json(&format!(
                "{}?f=get_inbox&lang={}&sid_token={}",
                BASE_URL, self.config.lang, sid
            ))
            .await?;

        let list = inbox.list.unwrap_or_default();
        Ok(list
            .into_iter()
            .map(|m| Message {
                id: m.mail_id,
                sender: m.mail_from,
                subject: m.mail_subject,
                date: parse_date(&m.mail_date), preview: String::new(), has_attachments: false })
            .collect())
    }

    async fn read_message(&self, message_id: &str) -> Result<MessageDetail, TempMailError> {
        let sid = self.session_id.lock().await;
        let sid = sid
            .as_ref()
            .ok_or_else(|| TempMailError::Api("Must call generate_email first".into()))?
            .clone();

        let resp: EmailBodyResponse = self
            .client
            .get_json(&format!(
                "{}?f=fetch_email&lang={}&sid_token={}&email_id={}",
                BASE_URL, self.config.lang, sid, message_id
            ))
            .await?;

        let (body_text, body_html) = match resp.content_type.as_deref() {
            Some("html") => (None, resp.mail_body),
            _ => (resp.mail_body, None),
        };

        Ok(MessageDetail {
            message: Message {
                id: resp.mail_id,
                sender: resp.mail_from,
                subject: resp.mail_subject,
                date: parse_date(&resp.mail_date), preview: String::new(), has_attachments: false },
            body_text,
            body_html,
            attachments: vec![], ..Default::default()         })
    }

    async fn delete_email(&self, email: &str) -> Result<bool, TempMailError> {
        let sid = self.session_id.lock().await;
        let sid = sid
            .as_ref()
            .ok_or_else(|| TempMailError::Api("Must call generate_email first".into()))?;

        #[derive(Deserialize)]
        struct DelResponse {
            is_disabled: Option<bool>,
        }

        let resp: DelResponse = self
            .client
            .get_json(&format!(
                "{}?f=del_email&email_ids[]={}&sid_token={}",
                BASE_URL, email, sid
            ))
            .await
            .unwrap_or(DelResponse {
                is_disabled: Some(false),
            });

        Ok(resp.is_disabled.unwrap_or(false))
    }
}
